Fixes to account for changes to the API of mu in version 1.7+

lordpretzel opened the pull request:
In mu version 1.7.x several internal functions have been renamed and/or behave differently. This pull requests adds explicit checks to use these new functions if `mu4e-mu-version` is larger than `1.7`.

xzz53 commented:
Looks good to me overall, just a few things to get it merged:

1. Fix byte-compiler warnings on both mu 1.6 and 1.7.
2. Too many `(version-list-<= '(1 7) (version-to-list mu4e-mu-version))` all over the place. Maybe `defconst` a variable like `mu4e-alert--mu-is-17` and use it instead?
3. Squash all the changes into a single commit

lordpretzel commented:
Hi,

I've refactored this to use `defalias` and `defvaralias` to avoid the constant if and introduced the `defconst` as suggested. This also fixes the byte compiler warnings. Plus I figured out that 1.7 does not directly set the `mu4e-header-func` and other variables which cases the `defvars` mu4e-alert uses to be `nil`. I've added a conditional call to the function that initializes these variables in 1.7. Let me know what you think.

BTW it's great to see that somebody is again maintaining this package!

xzz53 commented on mu4e-alert.el:
```diff
@@ -45,12 +45,41 @@
 (require 'pcase)
 (require 'cl-lib)
 
+;; deal with mu4e versions above 1.7
```
Let's put 1.7-related code below "Customizations" sections, so that `defcustom`s are easy to locate

xzz53 commented on mu4e-alert.el:
```diff
@@ -45,12 +45,41 @@
 
+;; make sure mu4e handlers are initialized
+(when mu4e-alert--mu-is-17
+  (mu4e--init-handlers))
```
This generates byte-compilation warning with mu 1.6

xzz53 commented on mu4e-alert.el:
```diff
@@ -45,12 +45,41 @@
+
+;; create aliases for mu4e functions
+(defalias 'mu4e-alert--mu4e-search
+  (if mu4e-alert--mu-is-17 'mu4e-search 'mu4e-headers-search)
```
`#'` should be used for function symbols

xzz53 commented on mu4e-alert.el:
```diff
@@ -45,12 +45,41 @@
+  "mu4e variables to store the current context.")
+
+(defalias 'mu4e-alert--mu4e-proc-running-p-func
+  (if mu4e-alert--mu-is-17 'mu4e-running-p 'mu4e~proc-running-p)
```
`#'`

xzz53 commented on mu4e-alert.el:
```diff
@@ -45,12 +45,41 @@
+  "Function that checks whether mu4e process is running.")
+
+(defalias 'mu4e-alert--mu4e-proc-find-func
+  (if mu4e-alert--mu-is-17 'mu4e--server-find 'mu4e~proc-find)
```
`#'`

xzz53 commented on mu4e-alert.el:
```diff
@@ -45,12 +45,41 @@
 
 (defgroup mu4e-alert nil
-  "Customization options for mu4e-alert"
+  "Customization options for mu4e-alert."
```
Let's leave this out for now. There are few more punctuation issues and I'm going to fix them myself in another commit.

xzz53 reviewed (changes requested):
Just a few more minor issues.

Note, that when you change `'` to `#'` as suggested, byte compiler warning will reappear. I believe there is no elegant solution for the 1.6->1.7 transition period, and we should just add `declare-function`s for all the changed functions (both old and new).

lordpretzel commented:
Hi,

so that gets us almost there. However, there is one remaining byte-compiler warning that one would get at 1.7 because `mu4e-headers-search` is marked as obsolete.

```
In toplevel form:
mu4e-alert.el:219:44: Warning: ‘mu4e-headers-search’ is an obsolete function (as of 1.7.0); use ‘mu4e-search’ instead.
```

The only way around that, that I know of is to remove the obsolete property:

```elisp
(eval-when-compile
  (cl-remprop 'mu4e-headers-search 'byte-obsolete-info))
```

However, this seems like a larger hack then just to accept the compiler warning in 1.7. What do you prefer?

xzz53 commented on mu4e-alert.el:
```diff
@@ -227,39 +263,41 @@ CALLBACK will be invoked by returned lambda"
                            #'mu4e-alert--get-mu-unread-emails-1
                            #'mu4e-alert--email-processor))
-      (setq mu4e-header-func 'mu4e-alert--header-func
+      (setq mu4e-alert-mu4e-header-func-var'mu4e-alert--header-func
```
whitespace before `'`

xzz53 commented on mu4e-alert.el:
```diff
@@ -189,7 +187,45 @@ See also https://github.com/jwiegley/alert."
+(declare-function mu4e~proc-find "mu4e" nil t)
+(declare-function mu4e--init-handlers "mu4e" nil t)
+(declare-function mu4e-search "mu4e-search" nil t)
+(eval-when-compile
```
MELPA policy is to have zero compilation warnings, so I propose to keep this and add a comment explaining the situation, so that this can be removed when mu 1.7 goes stable.

xzz53 reviewed (commented):
Hi,

and thanks for your effort! Looks like we are almost there. Beside two trivial changes I marked inline, please replace 
```diff
-(advice-add #'mu4e~proc-find :before #'mu4e-alert--find-start)
-(advice-add #'mu4e~headers-found-handler :before #'mu4e-alert--find-end)
+(advice-add #'mu4e-alert--mu4e-proc-find-func :before #'mu4e-alert--find-start)
+(advice-add 'mu4e-alert--found-func-save :before #'mu4e-alert--find-end)
```
It's preferred that you rebase your PR against master and have a single commit in the branch, so that it can be merged as is and we have a clean commit history.

lordpretzel committed:
account for changes in mu 1.7's API

xzz53 commented:
Merged, thanks!
